### PR TITLE
[Security] Add a SPIFFE ID parser

### PR DESCRIPTION
**Pull request by gtcooke94**
This adds a utility for parsing SPIFFE IDs from strings.

See the gRFC for more detail https://github.com/grpc/proposal/pull/462

### CODE REVIEW

File: src/core/credentials/transport/tls/spiffe_utils.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+#include <string>
+
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+
+namespace grpc_core {
+namespace experimental {
+
+class SpiffeId {
```
Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.h:30]: Please add comments on the class and methods.

File: src/core/credentials/transport/tls/spiffe_utils.cc [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_split.h"
+
+constexpr absl::string_view SPIFFE_PREFIX = "spiffe://";
+
+namespace grpc_core {
+namespace experimental {
+namespace {
+
+absl::Status doInitialUriValidation(absl::string_view uri) {
```
Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.cc:34]: nit: C++ naming should be `DoInitialUriValidation`.

Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.cc:34]: Please add a comment explaining what "initial URI validation" means.

File: src/core/credentials/transport/tls/spiffe_utils.cc [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        "SPIFFE ID cannot contain query fragments");
+  }
+  if (absl::StrContains(uri, "?")) {
+    return absl::InvalidArgumentError(
+        "SPIFFE ID cannot contain query parameters");
+  }
+  return absl::OkStatus();
+}
+
+absl::Status validateTrustDomain(absl::string_view trust_domain) {
```
Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.cc:55]: nit: C++ naming should be `ValidateTrustDomain`.

File: src/core/credentials/transport/tls/spiffe_utils.cc [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    if (c == '_') continue;
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "Trust domain contains invalid character %c. MUST contain only "
+        "lowercase letters, numbers, dots, dashes, and underscores",
+        c));
+  }
+  return absl::OkStatus();
+}
+
+absl::Status validatePathSegment(absl::string_view path_segment) {
```
Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.cc:77]: nit: C++ naming should be `ValidatePathSegment`.

File: src/core/credentials/transport/tls/spiffe_utils.cc [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    if (c == '_') continue;
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "Path segment contains invalid character %c. MUST contain only "
+        "letters, numbers, dots, dashes, and underscores",
+        c));
+  }
+  return absl::OkStatus();
+}
+
+absl::Status validatePath(absl::string_view path) {
```
Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.cc:100]: nit: C++ naming should be `ValidatePath`.

File: src/core/credentials/transport/tls/spiffe_utils.cc [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    if (absl::Status status = validatePathSegment(segment); !status.ok()) {
+      return status;
+    }
+  }
+  return absl::OkStatus();
+}
+
+}  // namespace
+
+// spiffe://trust-domain-name/path
```
Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.cc:114]: nit: Please remove.

File: src/core/credentials/transport/tls/spiffe_utils.cc [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+absl::StatusOr<SpiffeId> SpiffeId::FromString(absl::string_view uri) {
+  if (absl::Status status = doInitialUriValidation(uri); !status.ok()) {
```
Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.cc:116]: It looks like gRPC now has the `GRPC_RETURN_IF_ERROR` macro. Could we use that here and throughout the class?

File: src/core/credentials/transport/tls/spiffe_utils.cc [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+constexpr absl::string_view SPIFFE_PREFIX = "spiffe://";
+
+namespace grpc_core {
+namespace experimental {
+namespace {
+
+absl::Status doInitialUriValidation(absl::string_view uri) {
+  if (uri.empty()) {
+    return absl::InvalidArgumentError(
+        "SpiffeId cannot be parsed from empty uri");
```
Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.cc:37]: nit: s/SpiffeId/SPIFFE ID

Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.cc:37]: nit: s/uri/URI.

File: src/core/credentials/transport/tls/spiffe_utils.cc [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+#include "src/core/credentials/transport/tls/spiffe_utils.h"
+
+#include <string>
+
+#include "absl/strings/match.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_split.h"
+
+constexpr absl::string_view SPIFFE_PREFIX = "spiffe://";
```
Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.cc:28]: nit: I believe this is the old style in gRPC for constants, but (I believe) new code is following the google3 style, so you'd call the variable `kSpiffePrefix`.

File: src/core/credentials/transport/tls/spiffe_utils.cc [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+        "SPIFFE ID cannot contain query parameters");
+  }
+  return absl::OkStatus();
+}
+
+absl::Status validateTrustDomain(absl::string_view trust_domain) {
+  if (trust_domain.empty()) {
+    return absl::InvalidArgumentError("Trust domain cannot be empty");
+  }
+  if (trust_domain.size() >= 256) {
```
Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.cc:59]: nit: Can we make 256 a constant?

File: src/core/credentials/transport/tls/spiffe_utils.cc [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+        "Path segment contains invalid character %c. MUST contain only "
+        "letters, numbers, dots, dashes, and underscores",
+        c));
+  }
+  return absl::OkStatus();
+}
+
+absl::Status validatePath(absl::string_view path) {
+  if (path.empty()) {
+    return absl::OkStatus();
```
Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.cc:102]: Is an empty path allowed?

File: src/core/credentials/transport/tls/spiffe_utils.cc [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+      return status;
+    }
+  }
+  return absl::OkStatus();
+}
+
+}  // namespace
+
+// spiffe://trust-domain-name/path
+absl::StatusOr<SpiffeId> SpiffeId::FromString(absl::string_view uri) {
```
Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.cc:115]: Can the first step be to confirm that all characters in the URI are ASCII? You can check this with something like:

```
for (char ch : uri) {
    if (!absl::ascii_isascii(ch)) {
      // error
    }
  }

```

This is a bug that has caught us in the past. Can you also please check the Java implementation to confirm this is done there? If not, can you please add it to the Java implementation?

File: test/core/credentials/transport/tls/spiffe_utils_test.cc [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_split.h"
+#include "absl/strings/string_view.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "test/core/test_util/test_config.h"
+
+using ::grpc_core::experimental::SpiffeId;
```
Comment (markdroth) [inline test/core/credentials/transport/tls/spiffe_utils_test.cc:34]: This shouldn't be needed once you remove the `experimental` namespace, as I suggested elsewhere.

File: test/core/credentials/transport/tls/spiffe_utils_test.cc [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+namespace grpc_core {
+namespace testing {
+
+using ::testing::TestWithParam;
+
+struct SpiffeIdFailureTestCase {
+  std::string test_name;
```
Comment (markdroth) [inline test/core/credentials/transport/tls/spiffe_utils_test.cc:42]: This field doesn't seem to be used anywhere.  I think it can be removed.

File: test/core/credentials/transport/tls/spiffe_utils_test.cc [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+TEST_P(SpiffeIdFailureTest, SpiffeIdTestFailure) {
+  const SpiffeIdFailureTestCase& test_case = GetParam();
+  absl::StatusOr<SpiffeId> spiffe_id =
+      SpiffeId::FromString(test_case.spiffe_id);
+  EXPECT_EQ(spiffe_id.status().code(), absl::StatusCode::kInvalidArgument);
+  EXPECT_THAT(spiffe_id.status().message(),
+              ::testing::HasSubstr(test_case.status_contains));
+}
+
+INSTANTIATE_TEST_SUITE_P(
```
Comment (markdroth) [inline test/core/credentials/transport/tls/spiffe_utils_test.cc:58]: I think this would be a lot more readable without trying to use parameterized tests; instead, we can just have a separate `TEST()` block for each test case.  It's true that it would repeat a bit of boilerplate for each test, but I don't think it's worth accepting that for improved readability.

Same thing below.

File: src/core/credentials/transport/tls/spiffe_utils.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (gtcooke94) [reply src/core/credentials/transport/tls/spiffe_utils.cc]: Done

Comment (gtcooke94) [reply src/core/credentials/transport/tls/spiffe_utils.cc]: Done

File: src/core/credentials/transport/tls/spiffe_utils.cc [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+#include "absl/strings/str_split.h"
+
+constexpr absl::string_view SPIFFE_PREFIX = "spiffe://";
+
+namespace grpc_core {
+namespace experimental {
+namespace {
+
+absl::Status doInitialUriValidation(absl::string_view uri) {
+  if (uri.empty()) {
```
[the remark this answers is not in the document]
Comment (gtcooke94) [reply src/core/credentials/transport/tls/spiffe_utils.cc:38]: The SPIFFE spec explicitly disallows % encoding: https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE-ID.md#21-trust-domain and https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE-ID.md#21-trust-domain (and the other part of a spiffe id is the hard-coded `spiffe://`)

Given that, do you still think we should use the URI library?

File: src/core/credentials/transport/tls/spiffe_utils.cc (hunk 11 above)
Comment (gtcooke94) [reply src/core/credentials/transport/tls/spiffe_utils.cc:111]: Yes: [Spec Section](https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE-ID.md#2-spiffe-identity)

> Valid SPIFFE IDs MUST have the scheme set to spiffe, include a non-zero trust domain name, and MUST NOT include a query or fragment component. In other words, a SPIFFE ID is defined in its entirety by the spiffe scheme and a site-specific hier-part which includes an authority component and an ***optional path***.

File: src/core/credentials/transport/tls/spiffe_utils.cc (hunk 12 above)
Comment (gtcooke94) [reply src/core/credentials/transport/tls/spiffe_utils.cc:123]: Added this to the `DoInitialUriValidation`

File: src/core/credentials/transport/tls/spiffe_utils.cc (hunk 7 above)
Comment (gtcooke94) [reply src/core/credentials/transport/tls/spiffe_utils.cc]: Done

File: src/core/credentials/transport/tls/spiffe_utils.h [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+#ifndef GRPC_SRC_CORE_CREDENTIALS_TRANSPORT_TLS_SPIFFE_UTILS_H
+#define GRPC_SRC_CORE_CREDENTIALS_TRANSPORT_TLS_SPIFFE_UTILS_H
+
+#include <string>
+
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+
+namespace grpc_core {
+namespace experimental {
```
[the remark this answers is not in the document]
Comment (gtcooke94) [reply src/core/credentials/transport/tls/spiffe_utils.h]: Removed

File: src/core/credentials/transport/tls/spiffe_utils.h (hunk 1 above)
Comment (gtcooke94) [reply src/core/credentials/transport/tls/spiffe_utils.h:31]: Done

File: src/core/credentials/transport/tls/spiffe_utils.h [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+
+class SpiffeId {
+ public:
+  absl::string_view trust_domain() { return trust_domain_; }
+  absl::string_view path() { return path_; }
+  static absl::StatusOr<SpiffeId> FromString(absl::string_view uri);
```
[the remark this answers is not in the document]
Comment (gtcooke94) [reply src/core/credentials/transport/tls/spiffe_utils.h]: Fixed

File: test/core/credentials/transport/tls/spiffe_utils_test.cc (hunk 13 above)
Comment (gtcooke94) [reply test/core/credentials/transport/tls/spiffe_utils_test.cc]: Removed

File: test/core/credentials/transport/tls/spiffe_utils_test.cc (hunk 14 above)
Comment (gtcooke94) [reply test/core/credentials/transport/tls/spiffe_utils_test.cc:40]: It's used in the `INSTANTIATE_TEST_SUITE_P` call to identify what each case is - in the failure case specifically it makes the test output more sensible for what the test is trying to fail

File: test/core/credentials/transport/tls/spiffe_utils_test.cc (hunk 15 above)
Comment (gtcooke94) [reply test/core/credentials/transport/tls/spiffe_utils_test.cc:56]: I started to swap to individual tests and I'll gently push back - I think it's nice to be able to see essentially a tabular view of each case and expected result. It makes it easy to glance and see full test coverage vs. having to hold what you've seen in your head as you scroll through test cases. e.g.

```
TEST(SpiffeIdFailures, Empty) {
  absl::StatusOr<SpiffeId> spiffe_id =
      SpiffeId::FromString(/*uri=*/"");
  EXPECT_EQ(spiffe_id.status().code(), absl::StatusCode::kInvalidArgument);
  EXPECT_THAT(spiffe_id.status().message(),
              ::testing::HasSubstr("empty uri"));
}

TEST(SpiffeIdFailures, TooLong) {
  absl::StatusOr<SpiffeId> spiffe_id =
      SpiffeId::FromString(/*uri=*/std::string(2049, 'a');
  EXPECT_EQ(spiffe_id.status().code(), absl::StatusCode::kInvalidArgument);
  EXPECT_THAT(spiffe_id.status().message(),
              ::testing::HasSubstr("maximum allowed for SPIFFE ID is 2048"));
}

...
```

If you feel strongly I'm happy to change it, just let me know

File: src/core/credentials/transport/tls/spiffe_utils.cc (hunk 16 above)
[the remark this answers is not in the document]
Comment (markdroth) [reply src/core/credentials/transport/tls/spiffe_utils.cc:38]: That is... insane.  Percent-encoding is an inherent part of URI processing, as per RFC-3986.  Why the hell are they using a URI format if they aren't going to accept perfectly valid URIs?

The URI library doesn't support *not* allowing percent-encoding, because why the hell would it?  So no, I guess we can't use the URI library here.

The SPIFFE standard is really incredibly dumb here.

File: src/core/credentials/transport/tls/spiffe_utils.cc [hunk 19]
```diff
@@ -111,19 +120,16 @@ absl::Status validatePath(absl::string_view path) {
...
 }  // namespace
 
-// spiffe://trust-domain-name/path
 absl::StatusOr<SpiffeId> SpiffeId::FromString(absl::string_view uri) {
-  if (absl::Status status = doInitialUriValidation(uri); !status.ok()) {
-    return status;
-  }
-  if (!absl::StartsWithIgnoreCase(uri, SPIFFE_PREFIX)) {
+  GRPC_RETURN_IF_ERROR(DoInitialUriValidation(uri));
+  if (!absl::StartsWithIgnoreCase(uri, kSpiffePrefix)) {
```
Comment (markdroth) [inline src/core/credentials/transport/tls/spiffe_utils.cc:125]: Why are we ignoring the case here?  I don't see anything in the SPIFFE spec that says we should accept other cases for the scheme.

File: src/core/credentials/transport/tls/spiffe_utils.cc [hunk 20]
```diff
@@ -111,19 +120,16 @@ absl::Status validatePath(absl::string_view path) {
...
     return absl::InvalidArgumentError("SPIFFE ID must start with spiffe://");
   }
   if (absl::EndsWith(uri, /*suffix=*/"/")) {
     return absl::InvalidArgumentError("SPIFFE ID cannot end with a /");
   }
   // The input definitely starts with spiffe://
-  absl::string_view trust_domain_and_path = uri.substr(SPIFFE_PREFIX.length());
+  absl::string_view trust_domain_and_path = uri.substr(kSpiffePrefix.length());
```
Comment (markdroth) [inline src/core/credentials/transport/tls/spiffe_utils.cc:132]: You can probably avoid this by using `absl::ConsumePrefix()` instead of `absl::StartsWith()` above.

File: src/core/credentials/transport/tls/spiffe_utils.h [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+#include <string>
+
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+
+namespace grpc_core {
+
+// A representation of a SPIFFE ID per the spec:
+// https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE-ID.md#the-spiffe-identity-and-verifiable-identity-document
+class SpiffeId {
```
Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.h:31]: nit: Should we mark the class as "final"?

File: test/core/credentials/transport/tls/spiffe_utils_test.cc [hunk 22]
```diff
@@ -0,0 +1,281 @@
...
+
+TEST(SpiffeId, TrustDomainTooLongFails) {
+  EXPECT_EQ(
+      SpiffeId::FromString(absl::StrCat("spiffe://", std::string(256, 'a')))
+          .status(),
+      absl::InvalidArgumentError(
+          "Trust domain maximum length is 255 characters"));
+}
+
+TEST(SpiffeId, TrustDomainInvalidCharacterFails) {
```
Comment (matthewstevenson88) [inline test/core/credentials/transport/tls/spiffe_utils_test.cc:89]: Depending on the implementation, this could be subtle, because `@` usually denotes the start of the user info of a URI's hostname. For that reason, could we (a) rename this test to `TrustDomainWithUserInfoFails`, (b) change the string to `spiffe://domain@userinfo` to make this more clear, and (c) add a different test where it fails due to an invalid character (e.g. using $ instead of @)?

File: test/core/credentials/transport/tls/spiffe_utils_test.cc [hunk 23]
```diff
@@ -0,0 +1,281 @@
...
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_split.h"
+#include "absl/strings/string_view.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "test/core/test_util/test_config.h"
+
+namespace grpc_core {
```
Comment (matthewstevenson88) [inline test/core/credentials/transport/tls/spiffe_utils_test.cc:34]: Please add the following test cases:
- Short scheme prefix: `spiffe:/trustdomain/path`
- Scheme with invalid character: `ſPiffe://trustdomain/path`
- Scheme prefix and trust domain missing: `://`
- Scheme missing prefix: `piffe://`
- Scheme missing suffix: `spiffe`
- Trust domain contains a non-ASCII character
- Trust domain contains a percent-encoded character
- Port is present in the trust domain: `spiffe://github.com:1234/path`
- Query parameters are present in the path: `spiffe://trustdomain/path/?query`
- Trust domain has a trailing slash: `spiffe://trustdomain/`
- Escaped character in the path: `spiffe://trustdomain/pat\h`
- Fragment present in the path: `spiffe://trustdomain/pa^h`
- Multiple slashes in the path: `spiffe://trustdomain/namespace//ns`
- Path contains a non-ASCII character: `spiffe://trustdomain/pſth`

Note: Some of the above test cases will hit code paths that are already covered by other test cases here, but the above test cases are probing different edge cases involved with URIs, and so are good to have as the implementation changes over time.

File: src/core/credentials/transport/tls/spiffe_utils.cc [hunk 24]
```diff
@@ -0,0 +1,159 @@
...
+  }
+  return absl::OkStatus();
+}
+
+absl::Status ValidatePath(absl::string_view path) {
+  if (path.empty()) {
+    return absl::OkStatus();
+  }
+  for (absl::string_view segment : absl::StrSplit(path, '/')) {
+    if (absl::Status status = ValidatePathSegment(segment); !status.ok()) {
```
Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.cc:114]: nit: Can we use `GRPC_RETURN_IF_ERROR` here?

File: src/core/credentials/transport/tls/spiffe_utils.h [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+ public:
+  // Build a SPIFFE ID from a string. Returns a bad status if the input cannot
```
Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.h:33]: Can we combine the two sentences to make it a bit more clear what this API does? For example:

Parses the input string as a SPIFFE ID, and returns an error status if the input string is not a valid SPIFFE ID.

File: src/core/credentials/transport/tls/spiffe_utils.h [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+  // be parsed to a valid SPIFFE ID.
+  static absl::StatusOr<SpiffeId> FromString(absl::string_view uri);
```
Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.h:35]: nit: Should we change the variable name to "input" since we do not know at the time of calling this API that it is a valid URI string?

File: test/core/credentials/transport/tls/spiffe_utils_test.cc (hunk 23 above)
Comment (matthewstevenson88) [inline test/core/credentials/transport/tls/spiffe_utils_test.cc:34]: In a follow-up CL, can you also set up fuzzing for the `Parse` API?

File: test/core/credentials/transport/tls/spiffe_utils_test.cc (hunk 22 above)
Comment (gtcooke94) [reply test/core/credentials/transport/tls/spiffe_utils_test.cc]: Done

File: test/core/credentials/transport/tls/spiffe_utils_test.cc (hunk 23 above)
Comment (gtcooke94) [reply test/core/credentials/transport/tls/spiffe_utils_test.cc:34]: Yes

Comment (gtcooke94) [reply test/core/credentials/transport/tls/spiffe_utils_test.cc:34]: Done, PTAL

Comment (matthewstevenson88) [reply test/core/credentials/transport/tls/spiffe_utils_test.cc:34]: Thanks!

File: src/core/credentials/transport/tls/spiffe_utils.cc [hunk 27]
```diff
@@ -0,0 +1,158 @@
...
+#include "src/core/util/status_helper.h"
+
+namespace grpc_core {
+namespace {
+
+constexpr absl::string_view kSpiffePrefix = "spiffe://";
+constexpr int kMaxTrustDomainLength = 255;
+
+// Checks broad conditions on the whole input before splitting into the
+// pieces of a SPIFFE id
```
Comment (matthewstevenson88) [inline src/core/credentials/transport/tls/spiffe_utils.cc:36]: nit: s/id/ID.

**Review by matthewstevenson88 — APPROVED** [review]

File: test/core/credentials/transport/tls/spiffe_utils_test.cc [hunk 28]
```diff
@@ -34,106 +34,331 @@
...
+
+TEST(SpiffeId, WeirdCapitalizationSuccess) {
+  absl::StatusOr<SpiffeId> spiffe_id =
+      SpiffeId::FromString("sPiffe://example.com/us");
+  ASSERT_TRUE(spiffe_id.ok()) << spiffe_id.status();
+  EXPECT_EQ(spiffe_id->trust_domain(), "example.com");
+  EXPECT_EQ(spiffe_id->path(), "/us");
+}
+
+TEST(SpiffeId, AllSpiffeCapitalizedSuccess) {
```
Comment (markdroth) [inline test/core/credentials/transport/tls/spiffe_utils_test.cc:251]: These next two tests seem like duplicates of the previous test.  Suggest removing.

File: test/core/credentials/transport/tls/spiffe_utils_test.cc [hunk 29]
```diff
@@ -34,106 +34,331 @@
...
+}
+
+TEST(SpiffeId, ContainsNonASCIIPathFails) {
+  EXPECT_EQ(SpiffeId::FromString("spiffe://foo.bar/fooµbar").status(),
+            absl::InvalidArgumentError(
+                "SPIFFE ID URI cannot contain non-ascii characters"));
+}
+
+TEST(SpiffeId, NoPathSuccess) {
+  absl::StatusOr<SpiffeId> spiffe_id =
```
Comment (markdroth) [inline test/core/credentials/transport/tls/spiffe_utils_test.cc:228]: Consider using `auto` here instead of `absl::StatusOr<SpiffeId>` to make the tests a little shorter.

File: test/core/credentials/transport/tls/spiffe_utils_test.cc [hunk 30]
```diff
@@ -34,106 +34,331 @@
...
+
+TEST(SpiffeId, LongPathSuccess) {
+  absl::StatusOr<SpiffeId> spiffe_id = SpiffeId::FromString(
+      "spiffe://example.com/country/us/state/FL/city/Miami");
+  ASSERT_TRUE(spiffe_id.ok()) << spiffe_id.status();
+  EXPECT_EQ(spiffe_id->trust_domain(), "example.com");
+  EXPECT_EQ(spiffe_id->path(), "/country/us/state/FL/city/Miami");
+}
+
+TEST(SpiffeId, TrustDomainDashesSuccess) {
```
Comment (markdroth) [inline test/core/credentials/transport/tls/spiffe_utils_test.cc:275]: I think you could combine almost all of the rest of these tests into a single one that covers all possible valid characters:

```
TEST(SpiffeId, AcceptedCharacters) {
  auto spiffe_id = SpiffeId::FromString(
      "spiffe://abcdefghijklmnopqrstuvwxyz1234567890.-_/abcdefghijklmnopqrstuvwxyz1234567890.-_");
  ASSERT_TRUE(spiffe_id.ok()) << spiffe_id.status();
  EXPECT_EQ(spiffe_id->trust_domain(), "abcdefghijklmnopqrstuvwxyz1234567890.-_");
  EXPECT_EQ(spiffe_id->path(), "/abcdefghijklmnopqrstuvwxyz1234567890.-_");
}
```

File: test/core/credentials/transport/tls/spiffe_utils_test.cc [hunk 31]
```diff
@@ -34,106 +34,331 @@
...
+      SpiffeId::FromString("ab?de").status(),
+      absl::InvalidArgumentError("SPIFFE ID cannot contain query parameters"));
+}
+
+TEST(SpiffeId, DoesNotStartWithSpiffeFails) {
+  EXPECT_EQ(SpiffeId::FromString("www://foo/bar").status(),
+            absl::InvalidArgumentError("SPIFFE ID must start with spiffe://"));
+}
+
+TEST(SpiffeId, ShortSchemePrefixFails) {
```
[the remark this answers is not in the document]
Comment (markdroth) [reply test/core/credentials/transport/tls/spiffe_utils_test.cc:66]: I'll let Matt make the call here, but in my opinion most of these duplicate tests don't actually provide any value.  I do agree with the principle that it can be useful to future-proof by testing edge cases that could trigger if the implementation changes in the future, but I think we should do that thoughtfully -- if we can imagine a reasonable implementation change that could trigger an edge case, then it's fine to add a test for it, but I don't think we should just add every single possible edge case we can think of.

Just as an example, if we already have a test for "string contains an invalid character", then I don't think we also need a test for "string contains *only* invalid characters" -- the latter seems like clearly a special case of the former, and I can't imagine any reasonable implementation that would break the latter without breaking the former.

In contrast, I could imagine a future implementation trying to parse the entire URI first and then look at the individual components, so it seems useful to add a test for "spiffe:path", to make sure that fails even though it's a valid URI.

[the remark this answers is not in the document]
Comment (matthewstevenson88) [reply test/core/credentials/transport/tls/spiffe_utils_test.cc:66]: Thanks Greg and Mark. Mark's point is valid - thanks for calling this out. For context, we have a large internal corpus of test SPIFFE IDs that have caused issues in one of the several SPIFFE parsers, and so I grabbed the test cases from this corpus without checking whether all of them were reasonable in the case of this specific parser. I'll reply to Mark's comments indicating whether I agree that some cases can be removed.

File: test/core/credentials/transport/tls/spiffe_utils_test.cc [hunk 32]
```diff
@@ -34,106 +34,331 @@
...
+            absl::InvalidArgumentError("SPIFFE ID must start with spiffe://"));
+}
+
+TEST(SpiffeId, SchemeInvalidCharacterFails) {
+  EXPECT_EQ(SpiffeId::FromString("ſPiffe://trustdomain/path").status(),
+            absl::InvalidArgumentError(
+                "SPIFFE ID URI cannot contain non-ascii characters"));
+}
+
+TEST(SpiffeId, SchemePrefixAndTrustDomainMissingFails) {
```
[the remark this answers is not in the document]
Comment (matthewstevenson88) [reply test/core/credentials/transport/tls/spiffe_utils_test.cc:78]: That's fair, we can remove these three.

File: test/core/credentials/transport/tls/spiffe_utils_test.cc [hunk 33]
```diff
@@ -34,106 +34,331 @@
...
+
+TEST(SpiffeId, PathSegmentBadCharacterFails) {
+  EXPECT_EQ(
+      SpiffeId::FromString("spiffe://example/path/foo.bar/foo@bar").status(),
+      absl::InvalidArgumentError(
+          "Path segment contains invalid character @. MUST contain only "
+          "letters, numbers, dots, dashes, and underscores"));
+}
+
+TEST(SpiffeId, ContainsNonASCIITrustDomainFails) {
```
[the remark this answers is not in the document]
Comment (matthewstevenson88) [reply test/core/credentials/transport/tls/spiffe_utils_test.cc:160]: Agreed I think we can remove this one.

File: test/core/credentials/transport/tls/spiffe_utils_test.cc (hunk 28 above)
Comment (matthewstevenson88) [reply test/core/credentials/transport/tls/spiffe_utils_test.cc:255]: Agreed we can remove this and the next one.

File: test/core/credentials/transport/tls/spiffe_utils_test.cc [hunk 34]
```diff
@@ -34,106 +34,331 @@
...
+                "Path segment contains invalid character ^. MUST contain only "
+                "letters, numbers, dots, dashes, and underscores"));
+}
+
+TEST(SpiffeId, MultipleSlashesInPathFails) {
+  EXPECT_EQ(SpiffeId::FromString("spiffe://foo/bar//baz").status(),
+            absl::InvalidArgumentError("Path segment cannot be empty"));
+}
+
+TEST(SpiffeId, NonASCIIPathFails) {
```
[the remark this answers is not in the document]
Comment (gtcooke94) [reply test/core/credentials/transport/tls/spiffe_utils_test.cc:220]: Done

File: test/core/credentials/transport/tls/spiffe_utils_test.cc (hunk 32 above)
[the remark this answers is not in the document]
Comment (markdroth) [reply test/core/credentials/transport/tls/spiffe_utils_test.cc]: You removed the first one, but not the next two.  We should also remove the `SchemeMissingSuffixFails` and `SchemeMissingSuffixFails` tests.

[the remark this answers is not in the document]
Comment (gtcooke94) [reply test/core/credentials/transport/tls/spiffe_utils_test.cc]: Sorry, should be fixed now

**Review by markdroth — APPROVED** [review]
Looks great!
